Tests/GRPCTests/MutualTLSTests.swift

```diff
@@ -0,0 +1,156 @@
+    let clientErrorExpectation = self.expectation(description: "client error")
+    clientErrorExpectation.isInverted = expectedOutcome == .success
+    clientErrorExpectation.assertForOverFulfill = false
+    let clientErrorDelegate = ErrorRecordingDelegate(expectation: clientErrorExpectation)
+    clientConfiguration.errorDelegate = clientErrorDelegate
+
+    let client = Echo_EchoClient(channel: ClientConnection(configuration: clientConfiguration))
+
+    // Make the call.
+    let call = client.get(.with { $0.text = "mumble" })
+
+    // Wait for side effects.
+    self.wait(for: [clientErrorExpectation, serverErrorExpectation], timeout: 1)
+
+    switch expectedOutcome {
+    case .success:
+      // Verify no errors.
+      XCTAssert(serverErrorDelegate.errors.isEmpty)
+      XCTAssert(clientErrorDelegate.errors.isEmpty)
+      // Verify response.
+      let response = try call.response.wait()
+      XCTAssertEqual(response.text, "Swift echo get: mumble")
+      let status = try call.status.wait()
+      XCTAssertEqual(status.code, .ok)
+    case .serverError:
+      // Verify handshake error.
+      guard case .handshakeFailed = serverErrorDelegate.errors.first as? NIOSSLError else {
+        XCTFail("Expected NIOSSLError.handshakeFailed")
+        return
+      }
+    case .clientError:
+      // Verify handshake error.
+      guard case .handshakeFailed = clientErrorDelegate.errors.first as? NIOSSLError else {
+        XCTFail("Expected NIOSSLError.handshakeFailed")
+        return
+      }
+    }
+  }
+
+  func test_ValidClientAndServerCerts_Success() throws {
+    let serverTLSConfiguration = GRPCTLSConfiguration.makeServerConfigurationBackedByNIOSSL(
+      certificateChain: [.certificate(SampleCertificate.server.certificate)],
+      privateKey: .privateKey(SamplePrivateKey.server),
+      trustRoots: .certificates([SampleCertificate.ca.certificate]),
+      certificateVerification: .noHostnameVerification
+    )
+    let clientTLSConfiguration = GRPCTLSConfiguration.makeClientConfigurationBackedByNIOSSL(
+      certificateChain: [.certificate(SampleCertificate.client.certificate)],
+      privateKey: .privateKey(SamplePrivateKey.client),
+      trustRoots: .certificates([SampleCertificate.ca.certificate]),
+      certificateVerification: .fullVerification
+    )
+    try self.performTestWith(serverTLSConfiguration, clientTLSConfiguration, expect: .success)
+  }
+
+  func test_noServerCert_ClientError() throws {
+    let clientTLSConfiguration = GRPCTLSConfiguration.makeClientConfigurationBackedByNIOSSL(
+      certificateChain: [.certificate(SampleCertificate.client.certificate)],
+      privateKey: .privateKey(SamplePrivateKey.client),
+      trustRoots: .certificates([SampleCertificate.ca.certificate]),
+      certificateVerification: .fullVerification
+    )
+    try self.performTestWith(nil, clientTLSConfiguration, expect: .clientError)
+  }
+
+  func test_noClientCert_ServerError() throws {
```

Comment:
Similar to above: this validates that a plaintext client cannot connect to a server configured with TLS. The more interesting test would be to configure the client to not present any certificates and for the server to expect mTLS.

Reply:
Oh, my logic here for testing mutual auth was to check that when both sides provide valid certs things work, then when either side doesn't that things don't work.

Specifically this test is I guess testing client-verification by the server or am I missing something.

Is the thing I'm missing that we should be using a TLS client but with no cert (cf. no TLS)?

Reply:
> Specifically this test is I guess testing client-verification by the server or am I missing something.

Yes but more specifically I think we want to test that the server verifies the certificate presented by the client.

At the moment because the client isn't configured to use TLS, the first bytes it sends will be the http/2 preface which the server will attempt to interpret as the SSL client hello, so the server will fail the handshake at the step rather than because the client doesn't present a valid certificate.

> Is the thing I'm missing that we should be using a TLS client but with no cert (cf. no TLS)?

Yes

Reply:
OK, I've updated the PR description, and the `makecert` script which had some limitations, and updated the tests to hopefully make more sense.